core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java

```diff
@@ -0,0 +1,179 @@
+    ) {
+        for (var location : locations.get(0)) {
+            var startRange = new EdgeRange<>(location.edge, location.offset, edgeToLength.apply(location.edge));
+            registerStep(startRange, null, 0, 0);
+        }
+        while (true) {
+            var step = queue.poll();
+            if (step == null)
+                return null;
+            if (hasReachedEnd(locations, step))
+                return buildResult(step);
+            for (var stopLocation : locations.get(step.nReachedSteps + 1))
+                if (step.range.edge.equals(stopLocation.edge) && step.range.start <= stopLocation.offset) {
+                    // Adds a new step precisely on the stop location. This ensures that we don't ignore the
+                    // distance between the start of the edge and the stop location
+                    var newRange = new EdgeRange<>(stopLocation.edge, step.range.start, stopLocation.offset);
+                    registerStep(newRange, step.prev, step.totalDistance, step.nReachedSteps + 1);
+                }
+            var edgeLength = edgeToLength.apply(step.range.edge);
+            if (step.range.end == edgeLength) {
+                // We reach the end of the edge: we visit neighbors
+                var lastNode = graph.incidentNodes(step.range.edge).nodeV();
+                var neighbors = graph.outEdges(lastNode);
+                for (var edge : neighbors) {
+                    registerStep(
+                            new EdgeRange<>(edge, 0, edgeToLength.apply(edge)),
+                            step,
+                            step.totalDistance,
+                            step.nReachedSteps
+                    );
+                }
+            } else {
+                // We don't reach the end of the edge (intermediate step): we add a new step until the end
+                var newRange = new EdgeRange<>(step.range.edge, step.range.end, edgeLength);
+                registerStep(newRange, step, step.totalDistance, step.nReachedSteps);
+            }
+        }
+    }
+
+    /** Runs the pathfinding, returning a path as a list of edge. */
+    private List<EdgeT> runPathfindingEdgesOnly(
+            List<Collection<EdgeLocation<EdgeT>>> locations
+    ) {
+        var locatedSteps = runPathfinding(locations);
+        if (locatedSteps == null)
+            return null;
+        return locatedSteps.stream()
+                .map(step -> step.edge)
+                .collect(Collectors.toList());
+    }
+
+    /** Returns true if the step has reached the end of the path */
+    private boolean hasReachedEnd(
+            List<Collection<EdgeLocation<EdgeT>>> locations,
+            Step<EdgeT> step
+    ) {
+        return step.nReachedSteps >= locations.size() - 1;
+    }
+
+    /** Builds the result, iterating over the previous steps and merging ranges */
+    private List<EdgeRange<EdgeT>> buildResult(Step<EdgeT> step) {
+        var invertedResult = new ArrayList<EdgeRange<EdgeT>>();
```

Comment:
you can use an arraydeque and add to the front to make iteration slightly easier (it has the same complexity / almost same runtime code)